server: simplify bootstrap via fatter init server

tbg opened the pull request:
## Motivation

Today, starting up a server is complicated. This is especially true
when bootstrap is necessary. By this, we mean that either

- no NodeID is known. This implies that none of the engines were
  initialized yet
- the NodeID is known (i.e. at least one store is initialized), but a
  new engine was added.

When the process first starts, and a NodeID is known, a ClusterID is
also known (since they get persisted together, on every store). For the
same reason, a persisted cluster version is known. In this case, we can
start a fully initialized cluster, so this is the easy case.

It is more difficult when no NodeID is known, in which case the server
is just starting up for the first time, with its engines all blank. It
needs to somehow allocate NodeIDs (and StoreIDs) which are then written
to the engines. It also needs to, at least initially, use the lowest
possible cluster version it can tolerate (after all, the cluster may
actually run at that lowest version). Right now, there is a delicate
dance: we thread late-bound ClusterID and NodeID containers all over the
place, spin up a mostly dysfunctional Node, use its KV client to
allocate NodeID and StoreIDs once this is possible - we need Gossip
to have connected first - and update the containers. It is complex,
error prone, and ossifies any code it touches.

Cluster versions deserve an extra shout-out for complexity. Even if
a cluster version was persisted, the node may have been decommissioned
and the cluster upgraded. Much work went into our RPC layer to prevent
connections between incompatible nodes, but there is no boot-time check
that results in a swift and descriptive error - there will be a fatal
error, originating from the RPC subsystem. One aim of our work is to
simplify this by checking the version via an RPC before setting too
many gears in motion.

## Context

This marks the beginning of a series of PRs aiming at improving the
startup code. Ultimately, the goal is to have the Node and all Stores
bootstrapped as early as possible, but in particular before starting
the KV or SQL server subsystems.

Furthermore, we want to achieve this without relying on Gossip, to
prepare for a split between the SQL and KV servers in the context of
multitenancy support (SQL server won't be able to rely on Gossip, but
will still need to announce itself to the KV servers).

## This PR

This PR is an initial simplifying refactor that can help achieve these
goals. The init server (which hosts the Bootstrap RPC) is given more
responsibilities: it is now directly in charge of determining which, if
any, engines are bootstrapped, and explicitly listens to Gossip as well
as the Bootstrap RPC. It returns the cluster ID to the main server
start-up goroutine when it is available. As a result, the main startup
code has simplified, and a thread to be pulled on further has appeared
and is called out in TODOs.

Down the road (i.e. in later PRs), the init server will bootstrap a
NodeID and StoreIDs even when joining an existing cluster. It will
initially mimic/front-load the strategy taken by Node today, i.e. use a
`kv.DB`, but ultimately will bypass Gossip completely and use a simple
RPC call to ask the existing cluster to assign these IDs as needed. This
RPC will also establish the active cluster version, which is required
for SQL multi-tenancy, and generally follows the ideas in #32574.

Release note: None

tbg committed:
server: fix buglet in TestClusterIDMismatch

It was mixing NodeIDs within a Node, which was exposed when I refactored
inspectEngines.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46843)
<!-- Reviewable:end -->

tbg committed:
server: simplify bootstrap via fatter init server

\## Motivation

Today, starting up a server is complicated. This is especially true
when bootstrap is necessary. By this, we mean that either

- no NodeID is known. This implies that none of the engines were
  initialized yet
- the NodeID is known (i.e. at least one store is initialized), but a
  new engine was added.

When the process first starts, and a NodeID is known, a ClusterID is
also known (since they get persisted together, on every store). For the
same reason, a persisted cluster version is known. In this case, we can
start a fully initialized cluster, so this is the easy case.

It is more difficult when no NodeID is known, in which case the server
is just starting up for the first time, with its engines all blank. It
needs to somehow allocate NodeIDs (and StoreIDs) which are then written
to the engines. It also needs to, at least initially, use the lowest
possible cluster version it can tolerate (after all, the cluster may
actually run at that lowest version). Right now, there is a delicate
dance: we thread late-bound ClusterID and NodeID containers all over the
place, spin up a mostly dysfunctional Node, use its KV client to
allocate NodeID and StoreIDs once this is possible - we need Gossip
to have connected first - and update the containers. It is complex,
error prone, and ossifies any code it touches.

Cluster versions deserve an extra shout-out for complexity. Even if
a cluster version was persisted, the node may have been decommissioned
and the cluster upgraded. Much work went into our RPC layer to prevent
connections between incompatible nodes, but there is no boot-time check
that results in a swift and descriptive error - there will be a fatal
error, originating from the RPC subsystem. One aim of our work is to
simplify this by checking the version via an RPC before setting too
many gears in motion.

\## Context

This marks the beginning of a series of PRs aiming at improving the
startup code. Ultimately, the goal is to have the Node and all Stores
bootstrapped as early as possible, but in particular before starting
the KV or SQL server subsystems.

Furthermore, we want to achieve this without relying on Gossip, to
prepare for a split between the SQL and KV servers in the context of
multitenancy support (SQL server won't be able to rely on Gossip, but
will still need to announce itself to the KV servers).

\## This PR

This PR is an initial simplifying refactor that can help achieve these
goals. The init server (which hosts the Bootstrap RPC) is given more
responsibilities: it is now directly in charge of determining which, if
any, engines are bootstrapped, and explicitly listens to Gossip as well
as the Bootstrap RPC. It returns the cluster ID to the main server
start-up goroutine when it is available. As a result, the main startup
code has simplified, and a thread to be pulled on further has appeared
and is called out in TODOs.

Down the road (i.e. in later PRs), the init server will bootstrap a
NodeID and StoreIDs even when joining an existing cluster. It will
initially mimic/front-load the strategy taken by Node today, i.e. use a
`kv.DB`, but ultimately will bypass Gossip completely and use a simple
RPC call to ask the existing cluster to assign these IDs as needed. This
RPC will also establish the active cluster version, which is required
for SQL multi-tenancy, and generally follows the ideas in #32574.

Release note: None

tbg committed:
server: clarify when Gossip is and needs to be connected

Pre these refactors, many places in the code had to care about
whether Gossip had already connected, which was used as a proxy for
determining whether the clusterID was available.

Now, though, we always have the clusterID before the Node even gets
started. As a result, while we *do* want to log when Gossip does
connect, internally it matters very little.

As a result, various code paths have been simplified. In particular,
the Node was awkwardly involved in verifying that we weren't
cross-connecting between clusters, which has now been pulled out.

After this commit, the few remaining uses of `gossip.Connected` are
sane.

Release note: None

tbg committed:
server: fix ReadyFn when waitForInit==true

ReadyFn was never called with waitForInit set to true since we were
already past blocking when we'd initiate the call.

Caught by `pkg/cli/interactive_tests/test_init_command.tcl`.

Release note: None

andreimatei reviewed (commented):
Please see the "so what's the idea here" comment.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-:-M4Evkbv2Peq9MhygR1w:b9qu7ti)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/server/config.go, line 265 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4ErcOp97akSMp4UtHM:-M4ErcOp97akSMp4UtHN:b-d0m0mi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/config.go#L265)):*
> ```Go
> 	// will be waiting for an `init` command or accept bootstrapping
> 	// from a joined node. It is set in an advisory fashion, that is,
> 	// should be used for logging output only.
> ```

Hmm what is this new sentence trying to tell me?
___
*[pkg/server/init.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EryVK5xQ_MWkDqt24:-M4EryVK5xQ_MWkDqt25:b-am8el9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L35)):*
> ```Go
> 
> // initServer manages the temporary init server used during
> // bootstrapping.
> ```

mind adding more words here?
___
*[pkg/server/init.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsKYd2aHlpFz-mhpl:-M4EsKYd2aHlpFz-mhpm:b-u0ppaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L48)):*
> ```Go
> 
> 	binaryVersion, binaryMinSupportedVersion       roachpb.Version
> 	bootstrapVersion                               clusterversion.ClusterVersion
> ```

would `roachpb.Version` be more appropriate here?
___
*[pkg/server/init.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsQBg2OZJkJCFYmt5:-M4EsQBg2OZJkJCFYmt6:bwpso4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L48)):*
> ```Go
> 
> 	binaryVersion, binaryMinSupportedVersion       roachpb.Version
> 	bootstrapVersion                               clusterversion.ClusterVersion
> ```

please give `bootstrapVersion` a comment`
___
*[pkg/server/init.go, line 112 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EspWu2Q9s8cjuT6LP:-M4EspWu2Q9s8cjuT6LQ:bmfydgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L112)):*
> ```Go
> }
> 
> // ServeAndWait sets up the initServer to accept Bootstrap requests (which will
> ```

I think there's something wrong with the first sentence; the "which" seems off.
___
*[pkg/server/init.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsxOPA8_IgGGZYz2F:-M4EsxOPA8_IgGGZYz2G:brelc2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L113)):*
> ```Go
> 
> // ServeAndWait sets up the initServer to accept Bootstrap requests (which will
> // block until then). It uses the provided engines and gossip to block until
> ```

Can you explain 
___
*[pkg/server/init.go, line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EtPjj71wOuiDMuQ_h:-M4EtPjj71wOuiDMuQ_i:bb8kc9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L146)):*
> ```Go
> 		// now. Without this return, we could have all nodes in the cluster
> 		// deadlocked on g.Connected below. For similar reasons, we can't ever
> 		// hope to initialize the newEngines below, for which we would need to
> ```

can you pls add comments to the method about this - how new engines are not initialized if some of the engines were initialized already?
___
*[pkg/server/init.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Et6qp2PIIN6eYEzlB:-M4Et6qp2PIIN6eYEzlC:b-dcrvil) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L155)):*
> ```Go
> 	}
> 
> 	log.Info(ctx, "no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.")
> ```

long line
___
*[pkg/server/init.go, line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Ev8SR9fhfthq2lZqU:-M4Ev8SR9fhfthq2lZqV:b-jwllpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L217)):*
> ```Go
> 	ctx context.Context, request *serverpb.BootstrapRequest,
> ) (response *serverpb.BootstrapResponse, err error) {
> 	<-s.bootstrapBlockCh // block until ServeAndWait() is active
> ```

so what's the idea here? How come we open the port / register this RPC service before we decide whether we actually want to accept a `Bootstrap` RPC? Can't we just register the service after inspecting the engines?
___
*[pkg/server/init.go, line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Ev7PJ9CXziBpx_tRS:-M4Ev7PJ9CXziBpx_tRT:b-dcrvil) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L223)):*
> ```Go
> 		return nil, err
> 	}
> 	state, err := bootstrapCluster(ctx, s.engs, s.bootstrapVersion, s.bootstrapZoneConfig, s.bootstrapSystemZoneConfig)
> ```

long line


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46843)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/server/init.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsxOPA8_IgGGZYz2F:-M4EwADPAFusYEtypJ1G:b7xv2no) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L113)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Can you explain 
</blockquote></details>

Can you comment more on what the deal with the engines is - that we check to see if any had been bootstrapped? And also comment about under which circumstances engines get bootstrapped by this method, and when they don't?

It seems to me that it'd bundling the checking of engines with this method is weird. Do you think we could lift that part outside of the init server, or at least outside of anything that has to do with `close(s.bootstrapBlockCh)`? 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! As we discussed in our 1:1, I am reacting to your comments over in #46975.

bors r=andreimatei

> Dismissed @andreimatei from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-:-M4JBDk71P_LEV8tZGiy:b9xdtnb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/server/init.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EryVK5xQ_MWkDqt24:-M4J7R0X1ZtViwAJKBQI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L35)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mind adding more words here?
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsKYd2aHlpFz-mhpl:-M4J8cKiCenrFsV3Jzb5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

would `roachpb.Version` be more appropriate here?
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsQBg2OZJkJCFYmt5:-M4J8i319EFagTeptH1k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please give `bootstrapVersion` a comment`
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EsxOPA8_IgGGZYz2F:-M4JAvvm2cfMm5JvDEuL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L113)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Can you comment more on what the deal with the engines is - that we check to see if any had been bootstrapped? And also comment about under which circumstances engines get bootstrapped by this method, and when they don't?

It seems to me that it'd bundling the checking of engines with this method is weird. Do you think we could lift that part outside of the init server, or at least outside of anything that has to do with `close(s.bootstrapBlockCh)`? 
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4EtPjj71wOuiDMuQ_h:-M4JAmbn54dZoM4QXRmG:b-vqck7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L146)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you pls add comments to the method about this - how new engines are not initialized if some of the engines were initialized already?
</blockquote></details>

Done. The comments will get clearer once I give the initServer a *kv.DB. So far the awkwardness that when joining an existing cluster we only wait for Gossip and let the `*Node` handle the rest remains. Not for long.
___
*[pkg/server/init.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Et6qp2PIIN6eYEzlB:-M4J9td77v1YiKWPgXjz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L155)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

long line
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Ev8SR9fhfthq2lZqU:-M4J9Y155ab59Xev7FCa:bgmfxpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so what's the idea here? How come we open the port / register this RPC service before we decide whether we actually want to accept a `Bootstrap` RPC? Can't we just register the service after inspecting the engines?
</blockquote></details>

I managed to address this over in #46975.
___
*[pkg/server/init.go, line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46843#-M4Ev7PJ9CXziBpx_tRS:-M4J9kYBC0J85LijwmR1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7811bce13c80bf227c74f49bba6b3e9f31a9bb3/pkg/server/init.go#L223)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

long line
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
